src/vcpkg/versions.cpp

```diff
@@ -50,93 +51,185 @@ namespace vcpkg::Versions
+    {
+        auto cur = s;
+        for (; Parse::ParserBase::is_ascii_digit(*cur); ++cur)
         {
-            ret.version_string.resize(build_found);
         }
-
-        auto prerelease_found = ret.version_string.find('-');
-        if (prerelease_found != std::string::npos)
+        const char ch = *cur;
+        if (Parse::ParserBase::is_lower_alpha(ch) || Parse::ParserBase::is_upper_alpha(ch) || ch == '-')
         {
-            ret.prerelease_string = ret.version_string.substr(prerelease_found + 1);
-            ret.identifiers = Strings::split(ret.prerelease_string, '.');
-            ret.version_string.resize(prerelease_found);
+            // matched alpha identifier
+            ++cur;
+            for (; Parse::ParserBase::is_alphanumdash(*cur); ++cur)
+            {
+            }
+            return cur;
         }
-
-        std::regex version_match("(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)){2}");
-        if (!std::regex_match(ret.version_string, version_match))
+        if (*s == '0')
         {
-            return Strings::format("Error: String `%s` does not follow the required MAJOR.MINOR.PATCH format.",
-                                   ret.version_string);
+            // matches exactly zero
+            return s + 1;
         }
+        if (cur != s)
+        {
+            // matched numeric sequence (not starting with zero)
+            return cur;
+        }
+        return nullptr;
+    }
+
+    ExpectedS<SemanticVersion> SemanticVersion::from_string(const std::string& str)
+    {
+        // Suggested regex by semver.org
+        // std::regex semver_scheme_match("^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)"
+        //                                "(?:-((?:0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9][0-9]*|[0-9]"
+        //                                "*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
+        //                                "(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$");
+
+        SemanticVersion ret;
+        ret.original_string = str;
+        do
+        {
+            // (0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)
+            ret.version.resize(3);
+            const char* cur = str.c_str();
+            cur = parse_skip_number(cur, &ret.version[0]);
+            if (!cur || *cur != '.') break;
+            cur = parse_skip_number(cur + 1, &ret.version[1]);
+            if (!cur || *cur != '.') break;
+            cur = parse_skip_number(cur + 1, &ret.version[2]);
+            if (!cur) break;
+            ret.version_string.assign(str.c_str(), cur);
+            if (*cur == 0) return ret;
+
+            // pre-release
+            if (*cur == '-')
+            {
+                ++cur;
+                const char* const start_of_prerelease = cur;
+                for (;;)
+                {
+                    const auto start_identifier = cur;
+                    cur = skip_prerelease_identifier(cur);
+                    if (!cur) break;
```

Comment:
Here's a great example that would be made better with return since it has the cascaded seemingly meaningless extra if on 170.